### PR TITLE
Rewrite RzReg to use RzBitVector + RzFloat 80-bit

### PR DESCRIPTION
**Pull request by wargio**
# DO NOT SQUASH

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

- Rewrite RzReg to use RzBitVector
- Implement 128-bit float

### CODE REVIEW

File: librz/util/float/float.c [hunk 1]
```diff
@@ -84,6 +122,55 @@ RZ_API RZ_OWN char *rz_float_as_string(RZ_NULLABLE RzFloat *f) {
 	return str;
 }
 
+/**
+ * \brief return a numeric string of the float
```
Comment (XVilka) [inline librz/util/float/float.c]: I suggest to provide an example in detailed description

File: librz/util/float/float.c [hunk 2]
```diff
@@ -301,6 +399,65 @@ RZ_API RZ_OWN RzFloat *rz_float_new_from_double(double value) {
 	return f;
 }
 
+RZ_API RZ_OWN RzFloat *rz_float_new_from_quadruple(long double value) {
```
Comment (XVilka) [inline librz/util/float/float.c]: Add Doxygen

**Review by XVilka — CHANGES_REQUESTED** [review]
Nice cleanup! Due to how this depends on the target platform, I suggest resending it from `dist-asan-fuzz-*` branch, though.

**Comment by XVilka** [comment]
582 broken tests - most of them are related to the output of the same value:
```
@@ -0,5 +0,5 @@
 mov rax, 4
 mov rbx, 5
-rax = 0x00000004
-rbx = 0x00000005
+rax = 0x0000000000000004
+rbx = 0x0000000000000005
```
```
@@ -1,4 +1,4 @@
-cf = 0x00000000
-sf = 0x00000001
-of = 0x00000000
-zf = 0x00000000
+cf = 0x0
+sf = 0x1
+of = 0x0
+zf = 0x0
```

There are also seems to be genuinely broken tests, like:
```
[XX] db/esil/arm_64 bic 64-bit register
RZ_NOPLUGINS=1 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-693079402/rizinorg/rizin/prefix/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e asm.arch=arm
e asm.bits=64
wx 4a012b8a #bic x10, x11
ar x10=0x1122334455667788
ar x11=0xff
aes
ar x10
' malloc://0x200
-- stdout
--- expected
+++ actual
@@ -1,1 +1,0 @@
-x10 = 0x1122334455667700

-- exit status: -1

[XX] db/esil/arm_64 movz, movk
RZ_NOPLUGINS=1 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-693079402/rizinorg/rizin/prefix/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e asm.arch=arm
e asm.bits=64
wx ea7bbd52caff9772eaceac722a2dcdf2
ar w10=0x11223344aabbccdd
aeip
aes
ar x10
aes
ar x10
aes
ar x10
aes
ar x10
' malloc://0x200
-- stdout
--- expected
+++ actual
@@ -1,4 +1,0 @@
-x10 = 0xebdf0000
-x10 = 0xebdfbffe
-x10 = 0x6677bffe
-x10 = 0x69696677bffe

-- exit status: -1

[XX] db/formats/dyldcache dyldcache CVE-2022-1244
RZ_NOPLUGINS=1 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-693079402/rizinorg/rizin/prefix/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'aaa
q
' bins/dyldcache/CVE-2022-1244
-- stderr
[x] Analyze all flags starting with sym. and entry0 (aa)
[x] Analyze function calls
[x] find and analyze function preludes
[x] Analyze len bytes of instructions for references
[x] Check for classes
[x] Finding xrefs in noncode section with analysis.in=io.maps
[x] Analyze value pointers (aav)
[x] Value from 0x00000000 to 0x000002b0 (aav)
[x] 0x00000000-0x000002b0 in 0x0-0x2b0 (aav)
[ ] Emulate functions to find computed references
-- exit status: -1

[XX] db/formats/dyldcache dyldcache arm64 with objc
RZ_NOPLUGINS=1 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-693079402/rizinorg/rizin/prefix/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'i
?e --
ic
icf Test
icm Test
icf __NSUnrecognizedTaggedPointer
icm __NSUnrecognizedTaggedPointer
?e --
pxr 8 @ section.lib_libhello_1.0.dylib.11.__DATA_CONST.__objc_classlist
' bins/dyldcache/dyld_shared_cache_arm64-macOS-objc
-- stdout
--- expected
+++ actual
@@ -73,4 +73,3 @@
 0x18003e8e0 1     __NSUnrecognizedTaggedPointer       release
 0x18003e8e4 2     __NSUnrecognizedTaggedPointer       autorelease
 --
-0x182064028 0x0000000182068068   h....... @ section.lib_libhello_1.0.dylib.11.__DATA_CONST.__objc_classlist 6476431464 lib_libhello-1.0.dylib.17.__DATA.__objc_data _OBJC_CLASS_$_Test R 0x182068040

-- exit status: -1

[XX] db/analysis/mips mips ld(load doubleword) instruction
RZ_NOPLUGINS=1 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-693079402/rizinorg/rizin/prefix/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 's 0x120004460
e asm.bytes=true
f-*
fs-*
pd 1
' bins/elf/analysis/mips64r2-busybox-loongson
-- stdout
--- expected
+++ actual
@@ -1,1 +1,0 @@
-            0x120004460      208084df       ld    a0, -0x7fe0(gp)      ; [0x1200ee290:8]=0x120005074

-- exit status: -1

[XX] db/analysis/java Crash on analysis
RZ_NOPLUGINS=1 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-693079402/rizinorg/rizin/prefix/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e scr.prompt=false ; aaa' bins/java/fba022731cfdd80bc65a899afefead1e
-- stderr
ERROR: java bin: invalid constant pool tag: 23 at 0x2ab
ERROR: java bin: could not parse the constant pool value at offset 2ab
WARNING: (../librz/bin/format/java/class_bin.c:212):java_class_parse: code should not be reached

-- exit status: -1
```

**Comment by Heersin** [comment]
> 

why does it affect the esil test ? or the esil tests are already broken?

**Comment by wargio** [comment]
> > 
> 
> why does it affect the esil test ? or the esil tests are already broken?

i just broke everything, but it does affect ESIL because i have removed the old RzReg code

**Review by XVilka — APPROVED** [review]
LGTM, but please resend from `dist-*` branch to trigger everything.

**Comment by XVilka** [comment]
@wargio Windows build is broken:
```
FAILED: librz/util/rz_util.lib.p/float_float.c.obj 
"cl" "-Ilibrz\util\rz_util.lib.p" "-I." "-I.." "-Ilibrz" "-I..\librz" "-Ilibrz\include" "-I..\librz\include" "-Ilibrz\util\sdb\src" "-I..\librz\util\sdb\src" "-Ilibrz\util\sdb" "-I..\librz\util\sdb" "-Isubprojects\zlib-1.2.12" "-I..\subprojects\zlib-1.2.12" "-I..\subprojects\xz-5.2.5\src\liblzma\api" "/MT" "/nologo" "/showIncludes" "/utf-8" "/W2" "/O2" "/Gw" "-DPSAPI_VERSION=1" "-DRZ_PLUGIN_INCORE=1" "-DLZMA_API_STATIC" "/Fdlibrz\util\rz_util.lib.p\float_float.c.pdb" /Folibrz/util/rz_util.lib.p/float_float.c.obj "/c" ../librz/util/float/float.c
../librz/util/float/float.c(454): error C2065: 'INFINITY': undeclared identifier
../librz/util/float/float.c(481): error C2065: 'INFINITY': undeclared identifier
../librz/util/float/float.c(510): error C2065: 'INFINITY': undeclared identifier
../librz/util/float/float.c(539): error C2065: 'INFINITY': undeclared identifier
../librz/util/float/float.c(2073): warning C4146: unary minus operator applied to unsigned type, result still unsigned
[186/1940] Compiling C object librz/util/rz_util.lib.p/calc.c.obj
```
```
FAILED: librz/util/rz_util-0.5.dll.p/float_float.c.obj 
"clang-cl" "-Ilibrz\util\rz_util-0.5.dll.p" "-I." "-I.." "-Ilibrz" "-I..\librz" "-Ilibrz\include" "-I..\librz\include" "-Ilibrz\util\sdb\src" "-I..\librz\util\sdb\src" "-Ilibrz\util\sdb" "-I..\librz\util\sdb" "-Isubprojects\zlib-1.2.12" "-I..\subprojects\zlib-1.2.12" "-I..\subprojects\xz-5.2.5\src\liblzma\api" "/MD" "/nologo" "/showIncludes" "/utf-8" "/W2" "/O2" "-DPSAPI_VERSION=1" "-fcommon" "-D__STDC__=1" "-D_CRT_DECLARE_NONSTDC_NAMES " "-D_CRT_SECURE_NO_WARNINGS" "-D_CRT_NONSTDC_NO_DEPRECATE" "-DRZ_PLUGIN_INCORE=1" "-Werror=sizeof-pointer-memaccess" "-DLZMA_API_STATIC" "/Fdlibrz\util\rz_util-0.5.dll.p\float_float.c.pdb" /Folibrz/util/rz_util-0.5.dll.p/float_float.c.obj "/c" ../librz/util/float/float.c
../librz/util/float/float.c(193,35): warning: format specifies type 'double' but the argument has type 'long double' [-Wformat]
        return rz_str_newf("%" LDBLFMTg, result);
                            ~~~~~~~~~~~  ^~~~~~
../librz/util/float/float.c(454,61): error: use of undeclared identifier 'INFINITY'
                return rz_float_new_inf(RZ_FLOAT_IEEE754_BIN_32, value != F32_PINF);
                                                                          ^
..\librz\include\rz_types_base.h(100,27): note: expanded from macro 'F32_PINF'
#define F32_PINF  ((float)INFINITY)
                          ^
../librz/util/float/float.c(481,61): error: use of undeclared identifier 'INFINITY'
                return rz_float_new_inf(RZ_FLOAT_IEEE754_BIN_64, value != F64_PINF);
                                                                          ^
..\librz\include\rz_types_base.h(103,20): note: expanded from macro 'F64_PINF'
#define F64_PINF  (INFINITY)
                   ^
../librz/util/float/float.c(510,61): error: use of undeclared identifier 'INFINITY'
                return rz_float_new_inf(RZ_FLOAT_IEEE754_BIN_80, value != F128_PINF);
                                                                          ^
..\librz\include\rz_types_base.h(106,33): note: expanded from macro 'F128_PINF'
#define F128_PINF ((long double)INFINITY)
                                ^
../librz/util/float/float.c(539,62): error: use of undeclared identifier 'INFINITY'
                return rz_float_new_inf(RZ_FLOAT_IEEE754_BIN_128, value != F128_PINF);
                                                                           ^
..\librz\include\rz_types_base.h(106,33): note: expanded from macro 'F128_PINF'
#define F128_PINF ((long double)INFINITY)
                                ^
1 warning and 4 errors generated.
```

**Comment by wargio** [comment]
> LGTM, but please resend from `dist-*` branch to trigger everything.

i will do it once i see it green here.

**Pull request closed without merging**